controllers/helmrelease_controller_release.go

```diff
@@ -0,0 +1,232 @@
+				return ctrl.Result{}, err
+			}
+		default:
+			if hr.Spec.GetUpgrade().GetRemediation().RetriesExhausted(hr) {
+				err = fmt.Errorf("upgrade retries exhausted")
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	// Load the chart
+	loadedChart, err := r.loadHelmChart(hc)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Record the release attempt
+	v2.HelmReleaseAttempted(hr, hc.GetArtifact().Revision, valuesChecksum)
+
+	// Perform the release
+	doRelease := run.Install
+	remediation := hr.Spec.GetInstall().GetRemediation()
+	successReason, failureReason := v2.InstallSucceededReason, v2.InstallFailedReason
+	if rel != nil {
+		doRelease = run.Upgrade
+		remediation = hr.Spec.GetUpgrade().GetRemediation()
+		successReason, failureReason = v2.UpgradeSucceededReason, v2.UpgradeFailedReason
+	}
+	rel, err = doRelease(*hr, loadedChart, values)
+	if err != nil {
+		v2.SetHelmReleaseCondition(hr, v2.ReleasedCondition, v1.ConditionFalse, failureReason, "")
+		remediation.IncrementFailureCount(hr)
+		r.event(*hr, hr.Status.LastAttemptedRevision, events.EventSeverityError,
+			fmt.Sprintf("Helm release failed: %s", err.Error()))
+		return ctrl.Result{}, err
+	}
+
+	// Register the post release values
+	v2.SetHelmReleaseCondition(hr, v2.ReleasedCondition, v1.ConditionTrue, successReason, "")
+	hr.Status.LastAppliedRevision = hr.Status.LastAttemptedRevision
+	hr.Status.LastReleaseRevision = util.ReleaseRevision(rel)
+	r.event(*hr, hr.Status.LastAttemptedRevision, events.EventSeverityInfo, fmt.Sprintf("Helm release succeeded"))
+
+	return ctrl.Result{RequeueAfter: hr.Spec.Interval.Duration}, nil
+}
+
+func (r *HelmReleaseReconciler) reconcileTest(ctx context.Context, run *runner.Runner, hr *v2.HelmRelease) (ctrl.Result, error) {
+	// Tests are not enabled for this release, return.
+	if !hr.Spec.GetTest().Enable {
+		return ctrl.Result{}, nil
+	}
+
+	// If we already run the test, return
+	if tested := meta.GetCondition(hr.Status.Conditions, v2.TestSuccessCondition); tested != nil {
+		return ctrl.Result{}, nil
+	}
+
+	// If the release did not succeed, return
+	released := meta.GetCondition(hr.Status.Conditions, v2.ReleasedCondition)
+	if released == nil || released.Status == v1.ConditionFalse {
+		return ctrl.Result{}, nil
+	}
+
+	if _, err := run.Test(*hr); err != nil {
+		v2.SetHelmReleaseCondition(hr, v2.TestSuccessCondition, v1.ConditionFalse, v2.TestFailedReason, "")
+		r.event(*hr, hr.Status.LastAttemptedRevision, events.EventSeverityError,
+			fmt.Sprintf("Helm test failed: %s", err.Error()))
+
+		remediation := hr.Spec.GetInstall().GetRemediation()
+		if released.Reason == v2.UpgradeSucceededReason {
+			remediation = hr.Spec.GetUpgrade().GetRemediation()
+		}
+
+		if remediation.MustIgnoreTestFailures(hr.Spec.GetTest().IgnoreFailures) {
+			return ctrl.Result{}, nil
+		}
+
+		remediation.IncrementFailureCount(hr)
+		v2.SetHelmReleaseCondition(hr, v2.ReleasedCondition, v1.ConditionFalse, v2.TestFailedReason, "")
```

Comment:
...so just remove this line, and the `Released` condition becomes a reliable (barring status update failures) indicator of the last install/upgrade attempt against the current state.

Reply:
I guess though if we remove the test result from the Released condition, and later implement https://github.com/fluxcd/flux2/discussions/103, then the initial test result could get overwritten by subsequent interval-based tests. So then in `reconcileRemediation` we wouldn't be able to distinguish between an initial test failure and an interval based one without adding some additional distinguishing bit of status. Perhaps this could be simply using an alternate `reason` in the `TestSuccess` condition, e.g.  `IntervalTest<Succeeded|Failed>` (vs just `Test<Succeeded|Failed>` for the initial test).